Remora.Commands/Builders/CommandParameterBuilder.cs

```diff
@@ -0,0 +1,479 @@
+    public CommandParameterBuilder WithType(Type type)
+    {
+        if (_defaultValue is not null && _defaultValue.GetType() != type)
+        {
+            throw new InvalidOperationException("The type of the parameter must match the default value's type.");
+        }
+
+        _parameterType = type;
+        return this;
+    }
+
+    /// <summary>
+    /// Sets the parameter to be greedy.
+    /// </summary>
+    /// <returns>The builder to chain calls with.</returns>
+    public CommandParameterBuilder IsGreedy()
+    {
+        _isGreedy = true;
+
+        return this;
+    }
+
+    /// <summary>
+    /// Sets the parameter to be a named switch.
+    /// </summary>
+    /// <param name="defaultValue">The default value of the switch.</param>
+    /// <param name="shortName">The short name (e.g. '-o') of the switch.</param>
+    /// <param name="longName">The long name (e.g. '--option') of the switch.</param>
+    /// <returns>The builder to chain calls with.</returns>
+    /// <remarks>
+    /// This method will set a default value for the parameter, which will
+    /// implicitly make it optional.
+    /// </remarks>
+    /// <exception cref="InvalidOperationException">Thrown if <paramref name="shortName"/>
+    /// and <paramref name="longName"/> are both <c>null</c>.</exception>
+    public CommandParameterBuilder IsSwitch(bool defaultValue, char? shortName = null, string? longName = null)
+    {
+        _defaultValue = defaultValue;
+        _parameterType = typeof(bool);
+
+        if (shortName is null && longName is null)
+        {
+            throw new InvalidOperationException($"Either {nameof(shortName)} or {nameof(longName)} must be specified.");
+        }
+
+        var canBeSwitch = !_attributes.Any(r => r is OptionAttribute);
+
+        if (!canBeSwitch)
+        {
+            throw new InvalidOperationException("A parameter marked as an option cannot be a switch.");
+        }
+
+        var attribute = shortName is not null
+            ? longName is not null
+                ? new SwitchAttribute(shortName.Value, longName)
+                : new SwitchAttribute(shortName.Value)
+            : new SwitchAttribute(longName!);
+
+        _attributes.Add(attribute);
+
+        return this;
+    }
+
+    /// <summary>
+    /// Marks the parameter as being a named option.
+    /// </summary>
+    /// <param name="shortName">The short name (e.g. '-o') of the option.</param>
+    /// <param name="longName">The long name (e.g. '--option') of the option.</param>
+    /// <returns>The builder to chain calls with.</returns>
+    /// <exception cref="InvalidOperationException">Thrown if <paramref name="shortName"/>
+    /// and <paramref name="longName"/> are both <c>null</c>.</exception>
+    public CommandParameterBuilder IsOption(char? shortName = null, string? longName = null)
+    {
+        if (shortName is null && longName is null)
```

Comment:
You could use `OneOf<char, string>` here to get compile-time safety.

Reply:
This works, however doesn't allow for short AND long names. Perhaps `OneOf<char, string, (char, string)>`?